test/Test_Editing.re

```diff
@@ -16,14 +16,52 @@ module Fresh = Language.IdTagged.FreshGrammar;
 let caret_char = "¦"; /* Note this is two bytes */
 let convex_char = "?";
 let concave_char = "~";
+let selection_char = "§"; /* Note this is two bytes */
+let caret_regexp = StringUtil.regexp(caret_char);
 
-let printer = (z: Zipper.t): string =>
-  Printer.of_zipper(
-    ~holes=convex_char,
-    ~concave_holes=concave_char,
-    ~caret=caret_char,
-    z,
-  );
+let selection_size = (z: Zipper.t): option(int) => {
+  (
+    switch (z.selection) {
+    | {mode: Buffer(_), _} => None
+    | {content, mode: Normal, focus: Right} =>
+      Some(- String.length(Printer.of_segment(~holes=convex_char, content)))
+    | {content, mode: Normal, focus: Left} =>
+      Some(String.length(Printer.of_segment(~holes=convex_char, content)))
+    }
+  )
+  |> Util.OptUtil.filter(size => size != 0) /* Don't show selection if it's empty */;
+};
+
+let printer = (z: Zipper.t): string => {
+  let serialized_with_caret =
+    Printer.of_zipper(
+      ~holes=convex_char,
+      ~concave_holes=concave_char,
+      ~caret=caret_char,
+      z,
+    );
+  let caret_index =
+    switch (StringUtil.search(caret_regexp, serialized_with_caret, 0)) {
+    | Some((idx, _)) => idx
+    | None =>
+      Alcotest.fail("Failed to find caret in: " ++ serialized_with_caret)
+    };
+
+  let selection_size = selection_size(z);
+  // Place selection character n chars away from the caret
+  switch (selection_size) {
+  | None => serialized_with_caret
+  | Some(size) =>
+    let selection_pos = caret_index + size;
+    String.sub(serialized_with_caret, 0, selection_pos)
+    ++ selection_char
+    ++ String.sub(
+         serialized_with_caret,
+         selection_pos,
+         String.length(serialized_with_caret) - selection_pos,
+       );
+  };
+};
```

Comment:
is there a reason why you didn't bake the selection printing logic into Printer? it seems generically useful

Reply:
Would you calculate it the way I am and then add selection_point to Zipper like caret_point? Mostly I'm afraid of making something wrong in the actual code vs the tests.

Reply:
i know i said to do it the way youre doing, which is still fine by me, it but to do it via measured, i think you just do: take the first and last pieces in the selection and look up their first/last in the measurements. use the focus to decide which is the caret. use this logic only when the selection is non-empty, still caret_point otherwise

Reply:
@disconcision I think I did it but would like your review on https://github.com/hazelgrove/hazel/pull/1781/commits/e50db3b88b39746bdaa0614f5441e62ef27af4c8